constantine-go/constantine.go

```diff
@@ -265,4 +266,377 @@ func (ctx EthKzgContext) VerifyBlobKzgProofBatchParallel(tp Threadpool, blobs []
 		return false, err
 	}
 	return true, nil
-}
\ No newline at end of file
+}
+
+// Constantine's SHA256 API
+type Sha256Context C.ctt_sha256_context
+
+func Sha256ContextNew() (ctx Sha256Context, err error) {
+	return ctx, nil
+}
+
+func (ctx *Sha256Context) Sha256ContextInit() (bool, error) {
+	C.ctt_sha256_init((*C.ctt_sha256_context)(ctx))
+	return true, nil
+}
+
+func (ctx *Sha256Context) Sha256ContextUpdate(data []byte) (bool, error) {
+	C.ctt_sha256_update((*C.ctt_sha256_context)(ctx),
+		(*C.byte)(unsafe.Pointer(&data[0])),
+		(C.ptrdiff_t)(len(data)),
+	)
+	return true, nil
+}
+
+func (ctx *Sha256Context) Sha256ContextFinish(data [32]byte) (bool, error) {
+	C.ctt_sha256_finish((*C.ctt_sha256_context)(ctx),
+		(*C.byte)(unsafe.Pointer(&data[0])),
+	)
+	return true, nil
+}
+
+func (ctx *Sha256Context) Sha256ContextClear() (bool, error) {
+	C.ctt_sha256_clear((*C.ctt_sha256_context)(ctx))
+	return true, nil
+}
+
+func Sha256Hash(digest *[32]byte, message []byte, clearMemory bool) (bool, error) {
+	C.ctt_sha256_hash((*C.byte)(unsafe.Pointer(digest)),
+		(*C.byte)(unsafe.Pointer(&message[0])),
+		(C.ptrdiff_t)(len(message)),
+		(C.ctt_bool)(clearMemory),
+	)
+	return true, nil
+}
+
+// Ethereum BLS signatures
+// -----------------------------------------------------
+
+type (
+	EthBlsSecKey    C.ctt_eth_bls_seckey
+	EthBlsPubKey    C.ctt_eth_bls_pubkey
+	EthBlsSignature C.ctt_eth_bls_signature
+)
+
+func (pub EthBlsPubKey) IsZero() (bool, error) {
+	status := C.ctt_eth_bls_pubkey_is_zero((*C.ctt_eth_bls_pubkey)(&pub))
+	return bool(status), nil
+}
+
+func (sec *EthBlsSecKey) Deserialize(src [32]byte) (bool, error) {
+	status := C.ctt_eth_bls_deserialize_seckey((*C.ctt_eth_bls_seckey)(sec),
+		(*C.byte)(unsafe.Pointer(&src[0])))
+	if status != C.cttCodecScalar_Success {
+		err := errors.New(
+			C.GoString(C.ctt_codec_scalar_status_to_string(status)),
+		)
+		return false, err
+	}
+	return true, nil
+}
+
+func (pub *EthBlsPubKey) DerivePubKey(sec EthBlsSecKey) (bool, error) {
+	C.ctt_eth_bls_derive_pubkey((*C.ctt_eth_bls_pubkey)(pub), (*C.ctt_eth_bls_seckey)(&sec))
+	return true, nil
+}
+
+func (pub *EthBlsPubKey) Verify(message []byte, sig EthBlsSignature) (bool, error) {
+	status := C.ctt_eth_bls_verify((*C.ctt_eth_bls_pubkey)(pub),
+		(*C.byte)(unsafe.Pointer(&message[0])),
+		(C.ptrdiff_t)(len(message)),
```

Comment:
To schedule for later, ptr+len should use ptr+size_t as per C convention, and we cast to Nim int.